File: src/MbtiEnterpriseSimilarity.App/Domain/DimensionWeights.cs
using System.Globalization;

namespace MbtiEnterpriseSimilarity.App.Domain;

public sealed record DimensionWeights(
    double Ne,
    double Ni,
    double Te,
    double Ti,
    double Se,
    double Si,
    double Fe,
    double Fi)
{
    private const double Epsilon = 1e-12;

    public static DimensionWeights Equal { get; } = Create(1, 1, 1, 1, 1, 1, 1, 1);

    public bool IsEqualProfile =>
        AlmostEqual(Ne, Equal.Ne) &&
        AlmostEqual(Ni, Equal.Ni) &&
        AlmostEqual(Te, Equal.Te) &&
        AlmostEqual(Ti, Equal.Ti) &&
        AlmostEqual(Se, Equal.Se) &&
        AlmostEqual(Si, Equal.Si) &&
        AlmostEqual(Fe, Equal.Fe) &&
        AlmostEqual(Fi, Equal.Fi);

    public static DimensionWeights Parse(string rawValue)
    {
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            throw new ArgumentException("--weights must not be empty.");
        }

        var values = CognitiveScores.Dimensions.ToDictionary(
            dimension => dimension,
            _ => 1d,
            StringComparer.OrdinalIgnoreCase);

        var tokens = rawValue.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        if (tokens.Length == 0)
        {
            throw new ArgumentException("--weights format is invalid.");
        }

        foreach (var token in tokens)
        {
            var split = token.Split('=', StringSplitOptions.TrimEntries);
            if (split.Length != 2)
            {
                throw new ArgumentException($"Invalid weight token '{token}'. Use format like Ne=1.2");
            }

            var dimension = NormalizeDimension(split[0]);

            if (!double.TryParse(split[1], NumberStyles.Float, CultureInfo.InvariantCulture, out var parsedValue))
            {
                throw new ArgumentException($"Weight for '{dimension}' is not numeric: '{split[1]}'.");
            }

            values[dimension] = parsedValue;
        }

        return Create(
            values["Ne"],
            values["Ni"],
            values["Te"],
            values["Ti"],
            values["Se"],
            values["Si"],
            values["Fe"],
            values["Fi"]);
    }

    public static DimensionWeights Create(
        double ne,
        double ni,
        double te,
        double ti,
        double se,
        double si,
        double fe,
        double fi)
    {
        ValidateWeight(ne, "Ne");
        ValidateWeight(ni, "Ni");
        ValidateWeight(te, "Te");
        ValidateWeight(ti, "Ti");
        ValidateWeight(se, "Se");
        ValidateWeight(si, "Si");
        ValidateWeight(fe, "Fe");
        ValidateWeight(fi, "Fi");

        var total = ne + ni + te + ti + se + si + fe + fi;
        if (total <= Epsilon)
        {
            throw new ArgumentException("At least one weight must be greater than 0.");
        }

        return new DimensionWeights(
            Ne: ne / total,
            Ni: ni / total,
            Te: te / total,
            Ti: ti / total,
            Se: se / total,
            Si: si / total,
            Fe: fe / total,
            Fi: fi / total);
    }

    public IReadOnlyDictionary<string, double> ToDictionary() => new Dictionary<string, double>
    {
        ["Ne"] = Ne,
        ["Ni"] = Ni,
        ["Te"] = Te,
        ["Ti"] = Ti,
        ["Se"] = Se,
        ["Si"] = Si,
        ["Fe"] = Fe,
        ["Fi"] = Fi
    };

    public string ToCompactDisplay() =>
        string.Join(", ", ToDictionary().Select(pair => $"{pair.Key}={pair.Value.ToString("F4", CultureInfo.InvariantCulture)}"));

    private static bool AlmostEqual(double left, double right) => Math.Abs(left - right) < Epsilon;

    private static string NormalizeDimension(string rawDimension)
    {
        if (string.IsNullOrWhiteSpace(rawDimension))
        {
            throw new ArgumentException("Weight dimension key is empty.");
        }

        var match = CognitiveScores.Dimensions
            .FirstOrDefault(dimension => dimension.Equals(rawDimension.Trim(), StringComparison.OrdinalIgnoreCase));

        return match ?? throw new ArgumentException(
            $"Unknown dimension '{rawDimension}'. Valid keys: {string.Join(", ", CognitiveScores.Dimensions)}");
    }

    private static void ValidateWeight(double value, string name)
    {
        if (double.IsNaN(value) || double.IsInfinity(value))
        {
            throw new ArgumentException($"Weight for '{name}' must be a finite number.");
        }

        if (value < 0)
        {
            throw new ArgumentException($"Weight for '{name}' must be >= 0.");
        }
    }
}

File: src/main/java/com/mbti/similarity/domain/DimensionWeights.java
package com.mbti.similarity.domain;

import com.mbti.similarity.util.NumberFormats;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public final class DimensionWeights {
    private static final double EPSILON = 1e-12;

    public static final DimensionWeights EQUAL = create(1, 1, 1, 1, 1, 1, 1, 1);

    private final double ne;
    private final double ni;
    private final double te;
    private final double ti;
    private final double se;
    private final double si;
    private final double fe;
    private final double fi;

    private DimensionWeights(
        double ne,
        double ni,
        double te,
        double ti,
        double se,
        double si,
        double fe,
        double fi
    ) {
        this.ne = ne;
        this.ni = ni;
        this.te = te;
        this.ti = ti;
        this.se = se;
        this.si = si;
        this.fe = fe;
        this.fi = fi;
    }

    public static DimensionWeights parse(String rawValue) {
        if (rawValue == null || rawValue.isBlank()) {
            throw new IllegalArgumentException("--weights must not be empty.");
        }

        Map<String, Double> values = new LinkedHashMap<>();
        for (String dimension : CognitiveScores.DIMENSIONS) {
            values.put(dimension, 1.0d);
        }

        String[] tokens = rawValue.split(",");
        boolean hasToken = false;
        for (String token : tokens) {
            String trimmedToken = token.trim();
            if (trimmedToken.isEmpty()) {
                continue;
            }

            hasToken = true;
            String[] pair = trimmedToken.split("=", 2);
            if (pair.length != 2) {
                throw new IllegalArgumentException(
                    "Invalid weight token '" + trimmedToken + "'. Use format like Ne=1.2."
                );
            }

            String dimension = normalizeDimension(pair[0]);
            double parsedValue;
            try {
                parsedValue = Double.parseDouble(pair[1].trim());
            } catch (NumberFormatException ex) {
                throw new IllegalArgumentException(
                    "Weight for '" + dimension + "' is not numeric: '" + pair[1].trim() + "'.",
                    ex
                );
            }

            values.put(dimension, parsedValue);
        }

        if (!hasToken) {
            throw new IllegalArgumentException("--weights format is invalid.");
        }

        return create(
            values.get("Ne"),
            values.get("Ni"),
            values.get("Te"),
            values.get("Ti"),
            values.get("Se"),
            values.get("Si"),
            values.get("Fe"),
            values.get("Fi")
        );
    }

    public static DimensionWeights create(
        double ne,
        double ni,
        double te,
        double ti,
        double se,
        double si,
        double fe,
        double fi
    ) {
        validateWeight(ne, "Ne");
        validateWeight(ni, "Ni");
        validateWeight(te, "Te");
        validateWeight(ti, "Ti");
        validateWeight(se, "Se");
        validateWeight(si, "Si");
        validateWeight(fe, "Fe");
        validateWeight(fi, "Fi");

        double total = ne + ni + te + ti + se + si + fe + fi;
        if (total <= EPSILON) {
            throw new IllegalArgumentException("At least one weight must be greater than 0.");
        }

        return new DimensionWeights(
            ne / total,
            ni / total,
            te / total,
            ti / total,
            se / total,
            si / total,
            fe / total,
            fi / total
        );
    }

    public double getNe() {
        return ne;
    }

    public double getNi() {
        return ni;
    }

    public double getTe() {
        return te;
    }

    public double getTi() {
        return ti;
    }

    public double getSe() {
        return se;
    }

    public double getSi() {
        return si;
    }

    public double getFe() {
        return fe;
    }

    public double getFi() {
        return fi;
    }

    public boolean isEqualProfile() {
        return almostEqual(ne, EQUAL.ne)
            && almostEqual(ni, EQUAL.ni)
            && almostEqual(te, EQUAL.te)
            && almostEqual(ti, EQUAL.ti)
            && almostEqual(se, EQUAL.se)
            && almostEqual(si, EQUAL.si)
            && almostEqual(fe, EQUAL.fe)
            && almostEqual(fi, EQUAL.fi);
    }

    public Map<String, Double> toMap() {
        Map<String, Double> values = new LinkedHashMap<>();
        values.put("Ne", ne);
        values.put("Ni", ni);
        values.put("Te", te);
        values.put("Ti", ti);
        values.put("Se", se);
        values.put("Si", si);
        values.put("Fe", fe);
        values.put("Fi", fi);
        return values;
    }

    public String toCompactDisplay() {
        StringBuilder builder = new StringBuilder();
        boolean first = true;
        for (Map.Entry<String, Double> entry : toMap().entrySet()) {
            if (!first) {
                builder.append(", ");
            }
            builder.append(entry.getKey())
                .append('=')
                .append(NumberFormats.formatFixed(entry.getValue(), 4));
            first = false;
        }
        return builder.toString();
    }

    private static String normalizeDimension(String rawDimension) {
        Objects.requireNonNull(rawDimension, "rawDimension");
        String candidate = rawDimension.trim();
        if (candidate.isEmpty()) {
            throw new IllegalArgumentException("Weight dimension key is empty.");
        }

        for (String dimension : CognitiveScores.DIMENSIONS) {
            if (dimension.equalsIgnoreCase(candidate)) {
                return dimension;
            }
        }

        throw new IllegalArgumentException(
            "Unknown dimension '" + candidate + "'. Valid keys: " + String.join(", ", CognitiveScores.DIMENSIONS)
        );
    }

    private static void validateWeight(double value, String name) {
        if (!Double.isFinite(value)) {
            throw new IllegalArgumentException("Weight for '" + name + "' must be a finite number.");
        }

        if (value < 0d) {
            throw new IllegalArgumentException("Weight for '" + name + "' must be >= 0.");
        }
    }

    private static boolean almostEqual(double left, double right) {
        return Math.abs(left - right) < EPSILON;
    }
}
